[JENKINS-67744] make agent availability help more flexible

mawinter69 opened the pull request:
each retention strategy now has it's own help file that is included automatically.

The old help files need to be kept for now as other plugins are still relying on them

See [JENKINS-67744](https://issues.jenkins-ci.org/browse/JENKINS-67744).

Currently nearly all implementations in plugins are for cloud purposes and do not even provide a descriptor so they are not presented in the UI at all.
What this will not fix is when plugins implement own extensions of node/slave and use in their jelly files the existing code with the hard coded link to the help for the retention strategy. One could create an own tag in lib/form for this eventually

### Proposed changelog entries

* Improve agent availability help.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://www.jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

mawinter69 committed:
[JENKINS-67744] make agent availability help more flexible

each retention strategy now has it's own help file that is included
automatically.

The old help files need to be kept for now as other plugins are still
relying on them

NotMyFault reviewed (commented):
The double line breaks seem a bit unneeded, because they clutter the in-paragraph sentences quite a bit.

mawinter69 committed:
Update core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help_de.properties

Co-authored-by: Alex <mc.cache@web.de>

mawinter69 commented:
> The double line breaks seem a bit unneeded, because they clutter the in-paragraph sentences quite a bit.

I didn't change the formatting of the texts compared to what was there in the html. But that doesn't mean it has to stay like it

mawinter69 committed:
update formatting

add line breaks breaks at end or property files

remove the double linebreaks in the help texts

NotMyFault commented:
> 

Yeah you already got it sorted, I found the couple of newlines weird to read through, especially the one breaking one paragraph.

Though it looks like GitHub broke the file encoding while I suggested the typo fix, sorry for that.

mawinter69 committed:
more text formatting

mawinter69 committed:
text formatting

mawinter69 committed:
fix encoding

oleg-nenashev commented on core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly:
```diff
@@ -0,0 +1,42 @@
+<!--
+The MIT License
+
+Copyright (c) 2004-2009, Sun Microsystems, Inc.
```
Needs fix

oleg-nenashev commented on core/src/main/resources/hudson/slaves/Messages_nl.properties:
```diff
@@ -0,0 +1,26 @@
+# The MIT License
+# 
+# Copyright (c) 2004-2009, Sun Microsystems, Inc., Kohsuke Kawaguchi, Simon Wiest
```
Needs fix

oleg-nenashev commented on core/src/main/resources/hudson/slaves/Messages_nl.properties:
```diff
@@ -0,0 +1,26 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+RetentionStrategy.Always.displayName=Hou deze slaafnode zoveel mogelijk beschikbaar
```
slaafnode is "Slave Node"

mawinter69 replied:
Replaced slaafnode with agent at all places.
Hope that this is a correct to use agent here in dutch

oleg-nenashev reviewed (dismissed):
"slaafnode" is "slave node". This is old terminology, and it must not be used in new logic

mawinter69 committed:
fix license and texts

replace in the newly added texts for NL the word slaafnode with agent

timja commented:
Would be good to provide testing instructions for PRs like this.

I've created a permanent agent here:
http://localhost:8080/computer/createItem

Clicked the help icon next to 'Availability'

I get:
![image](https://user-images.githubusercontent.com/21194782/155953602-b4a6e911-b579-4baa-ba5b-8e6bf6f9348e.png)

Would it be better to do this how pipeline syntax does it:

![image](https://user-images.githubusercontent.com/21194782/155953699-5a288748-bf01-4dc6-b57a-3fec5c7278b5.png)

That way we don't have a huge help blob once plugins implement this?

Each one you choose can have more information added to it.

mawinter69 commented:
That is probably even better. Need to check how this is done in pipeline.
I looked up some of the other retention strategies and all were related to cloud agents and had no descriptor, so they wouldn't appear at all.

timja commented:
> I looked up some of the other retention strategies and all were related to cloud agents and had no descriptor, so they wouldn't appear at all.

Seems the Azure ones do 😂 .

mawinter69 commented:
https://github.com/jenkinsci/jenkins/pull/5417
actually contains parts of the solution to use `f:dropdownDescriptorSelector` 
Then only the help files would be required

mawinter69 committed:
use f:dropdownDescriptorSelector

add a help field that displays the help of the selected entry when
available

mawinter69 commented:
similar change could be done for the launch method

mawinter69 commented:
The change to the f:dropdownDescriptorSelector will have a negative impact to the ssh build agent plugin. There the help is manually added to each config.jelly file for the host key verification strategy.
There might be other places with the same problem

timja commented:
> The change to the f:dropdownDescriptorSelector will have a negative impact to the ssh build agent plugin. There the help is manually added to each config.jelly file for the host key verification strategy. There might be other places with the same problem

Link?

PRs can be filed there to fix it, especially if it was working around a core limitation

mawinter69 commented:
in the config.jelly files of the verifiers in https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers

timja commented:
> in the config.jelly files of the verifiers in [https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers](https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers?rgh-link-date=2022-02-28T15%3A05%3A30Z)

it documents it as a hack:
https://github.com/jenkinsci/ssh-slaves-plugin/blob/f3105af5e5d3adf9343c74ebae03c08b6c2fafe4/src/main/resources/hudson/plugins/sshslaves/verifiers/ManuallyProvidedKeyVerificationStrategy/config.jelly#L27

Could you send a PR to remove that?

timja commented:
The help icon is orphaned on it's own here:

<img width="1013" alt="image" src="https://user-images.githubusercontent.com/21194782/156059612-80f4436e-b8a1-4a58-9744-fdba794a0590.png">

I think it needs the display name or some other label there otherwise it looks really strange

mawinter69 commented:
Indeed it doesn't look nice, but repeating what was just written in the dropdown is also not so nice. I would have said put the question mark for the help to the right of the drop down, but this would make the looks inconsistent with other drop downs probably.

timja commented:
fyi https://github.com/jenkinsci/jenkins/pull/5417 is merged. could you merge in master?

mawinter69 committed:
Merge branch 'master' into JENKINS-67744

mawinter69 committed:
set class on forms

so the help button is properly placed

timja reviewed (approved):
nice work!

NotMyFault commented on core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly:
```diff
@@ -0,0 +1,32 @@
+-->
+
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
```
```suggestion
<j:jelly xmlns:j="jelly:core" xmlns:l="/lib/layout">
```

NotMyFault commented on core/src/main/resources/hudson/slaves/RetentionStrategy/Always/help.jelly:
```diff
@@ -0,0 +1,4 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
```
```suggestion
<j:jelly xmlns:j="jelly:core">
```

NotMyFault commented on core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help.jelly:
```diff
@@ -0,0 +1,4 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
```
```suggestion
<j:jelly xmlns:j="jelly:core">
```

NotMyFault commented on core/src/main/resources/hudson/slaves/SimpleScheduledRetentionStrategy/help.jelly:
```diff
@@ -0,0 +1,4 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
```
```suggestion
<j:jelly xmlns:j="jelly:core">
```

NotMyFault reviewed (commented):
Looks much better now

timja commented:
Hmm it looks like this highlights an issue similar to https://github.com/jenkinsci/jenkins/pull/4562

Azure retention strategy items are showing up in the help whereas they didn't before this PR:

<img width="597" alt="image" src="https://user-images.githubusercontent.com/21194782/162636477-a0fbe97b-0845-4165-8064-2689c4a94ebf.png">


(on 2.342)
<img width="568" alt="image" src="https://user-images.githubusercontent.com/21194782/162636512-acc67681-d834-469a-9b5e-59a8237afc4a.png">

It's using a `DescriptorVisibilityFilter` to hide it:

https://github.com/jenkinsci/azure-vm-agents-plugin/blob/master/src/main/java/com/microsoft/azure/vmagent/AzureVMCloudPoolRetentionStrategy.java#L188-L194

I've checked both the Kubernetes plugin and EC2 plugin.

Kubernetes plugin seems to have created it's own retention strategy not based off the core one.
EC2 plugin doesn't seem to use it either, there's a class for it but seems like all the options are available as part of the template config, which kinda makes sense there was a PR here which did something similar except just duplicated a lot of options rather than flattening it: https://github.com/jenkinsci/azure-vm-agents-plugin/pull/183

----

Not sure if we should proceed right now, anyone got any thoughts?

mawinter69 committed:
Update core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly

Co-authored-by: Alexander Brandes <brandes.alexander@web.de>

mawinter69 committed:
Update core/src/main/resources/hudson/slaves/RetentionStrategy/Always/help.jelly

Co-authored-by: Alexander Brandes <brandes.alexander@web.de>

mawinter69 committed:
Update core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help.jelly

Co-authored-by: Alexander Brandes <brandes.alexander@web.de>

mawinter69 committed:
Update core/src/main/resources/hudson/slaves/SimpleScheduledRetentionStrategy/help.jelly

Co-authored-by: Alexander Brandes <brandes.alexander@web.de>

mawinter69 committed:
ensure visibility filter is applied

mawinter69 commented:
I think I fixed this now 
Before it was looping over the filtered list, so I just made it the same.
EC2 plugin has no Extension on their descriptor so it is not shown.
Same for other retentionstrategies.

timja commented:
List is empty for me on the configure agent screen of a permanent agent:

![image](https://user-images.githubusercontent.com/21194782/162680698-682f1283-7123-499e-8793-cfa9a8e7d91c.png)

timja committed:
Provide the descriptor list

timja commented:
Fixed for you in https://github.com/jenkinsci/jenkins/pull/6255/commits/eec7a460eeca4ed647cf0ff89297936f5f09fe0b

![image](https://user-images.githubusercontent.com/21194782/162681072-e8cfb9d9-6b3a-4562-bbf0-f47777d6f5dd.png)

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
